### ISSUE TITLE
limbs_count() for Integer

### ISSUE THREAD

**Issue by amamic1803** [mhogrefe/malachite#33]
Hi,
I need to access limbs of Natural and twos complement limbs of Integer and process them. I suppose it is more efficient with iterator method instead of to, or into methods as I am processing limbs sequentially (don't need Vec). The problem is I need to know how many limbs there are before going through iterator. For Natural there is simple limbs_count method that just returns the len of internal vector. But there is no such method for Integer. I understand that it is much more complex operation for Integer since, depending on the highest bit value, additional limb must be used because of twos complement representation (am I right?). Still that seems like a simple if check and plus 1 in some potential twos_complement_limbs_count() method.
Is there any reason for such method not to exist?

P.S. I asked a few days ago about integration with PyO3. I decided conversion code should just be integrated into PyO3 directly instead of malachite itself and I'm working on that. This twos_complement_limbs_count method would be beneficial because right now I'm just using iterator.count().

**Comment by mhogrefe** [comment]
Hi,

Malachite uses a signed-magnitude representation for `Integer`s. (The code is [here](https://docs.rs/malachite-nz/latest/src/malachite_nz/integer/mod.rs.html#15-19).) The two's complement iterator has to do a bit of work to produce the twos' complement limbs, because they don't actually exist in memory. You can get a reference to the absolute value of an `Integer` using [`unsigned_abs_ref`](https://docs.rs/malachite-nz/latest/malachite_nz/integer/struct.Integer.html#method.unsigned_abs_ref), and then you can call `limbs_count` on that. So I recommend sticking with a signed-magnitude representation for PyO3.

But if you really want to count the two's complement limbs of an `Integer`, you can count the limbs of the absolute value, and add 1 (as you suggested) in these two cases:
- If the `Integer` is positive and the highest bit of the highest limb is set
- If the `Integer` is negative, the highest bit of the highest limb (of the absolute value) is set, and the remaining bits are not all zero (that is, the integer is not the negative of a power of 2). The code would look like this:

```rust
use malachite::integer::Integer;
use malachite::num::arithmetic::traits::{DivisibleByPowerOf2, IsPowerOf2};
use malachite::num::basic::integers::PrimitiveInt;
use malachite::num::logic::traits::SignificantBits;
use malachite::platform::Limb;

pub fn twos_complement_limb_count(n: &Integer) -> u64 {
    let abs_limbs_count = n.unsigned_abs_ref().limb_count();
    let highest_bit_of_highest_limb = n
        .significant_bits()
        .divisible_by_power_of_2(Limb::LOG_WIDTH);
    if highest_bit_of_highest_limb && (*n > 0 || (*n < 0 && !n.unsigned_abs_ref().is_power_of_2()))
    {
        abs_limbs_count + 1
    } else {
        abs_limbs_count
    }
}
```

Maybe I'll add this function in a future release.

**Comment by amamic1803** [comment]
Thank you for fast response! 😃
I need to use twos complement to convert to and from Python int type. If I didn't, I could probably pass the value around as a String but that would be less efficient. This way I can get value from Python as twos complement and send it the value as twos complement.
I knew much of the stuff you wrote about before (I only missed the case of integer being the negative of a power of 2) and I thank you for writing me a function for that. I wrote some tests for the function and it works correctly.

```rust
/// Fibonacci sequence iterator (Rust)
fn rust_fib<T>() -> impl Iterator<Item = T>
where
    T: From<u8>,
    for<'a> &'a T: std::ops::Add<Output = T>,
{
    let mut f0: T = T::from(1);
    let mut f1: T = T::from(1);
    std::iter::from_fn(move || {
        let f2 = &f0 + &f1;
        Some(std::mem::replace(&mut f0, std::mem::replace(&mut f1, f2)))
    })
}


/// Test twos_complement_limb_count function
#[test]
fn twos_limb_count() {
    // assert that the number of limbs is zero for zero
    assert_eq!(0, twos_complement_limb_count(&Integer::ZERO));
    assert_eq!(0, twos_complement_limb_count(&-Integer::ZERO));

    // test for the first 5000 fibonacci numbers
    for int_fib in rust_fib::<Integer>().take(5000) {
        // assert that the number of limbs is correct
        assert_eq!(twos_complement_limb_count(&int_fib), int_fib.twos_complement_limbs().count() as u64);

        // negate the number and assert that the number of limbs is correct
        let neg_int_fib = -int_fib;
        assert_eq!(twos_complement_limb_count(&neg_int_fib), neg_int_fib.twos_complement_limbs().count() as u64);
    }

    // test for first 5000 powers of 2
    for i in 0..5000 {
        // generate the power of 2
        let int_pow2 = Integer::from(1) << i;

        // assert that the number of limbs is correct
        assert_eq!(twos_complement_limb_count(&int_pow2), int_pow2.twos_complement_limbs().count() as u64);

        // negate the number and assert that the number of limbs is correct
        let neg_int_pow2 = -int_pow2;
        assert_eq!(twos_complement_limb_count(&neg_int_pow2), neg_int_pow2.twos_complement_limbs().count() as u64);
    }
}
```

If you decide to implement this function in the next version of malachite, please let me know to wait with my PR to PyO3 until the new version.

Now, I have another question, I've seen that malachite-nz crate has 32_bit_limbs feature flag to set limb size to 32 bits, however top level malachite crate doesn't have that feature flag. Is this supposed to be like that? I would like to support both 32 and 64 bit limbs, but malachite crate would need to export that 32_bit_limbs flag from malachite-nz for that.

Thanks for all the help!

**Comment by mhogrefe** [comment]
That makes sense. Ok, I'll implement it in the next version of Malachite. The missing 32_bit_limbs at the top level is just an oversight; I'll add it too.

**Comment by amamic1803** [comment]
I'll wait for new version then. No need to hurry, I won't have time to work on this the next few weeks anyway.
Thanks!

**Comment by mhogrefe** [comment]
Released as part of Malachite 0.4.5.

https://docs.rs/malachite-nz/latest/malachite_nz/integer/struct.Integer.html#method.twos_complement_limb_count

**Comment by amamic1803** [comment]
Thank you!
